File: iosFood2Fork/iosFood2Fork/presentation/recipe_detail/RecipeDetailViewModel.swift
//
//  RecipeDetailViewModel.swift
//  iosFood2Fork
//
//  Created by Vusi Moyo on 2021/10/16.
//  Copyright © 2021 orgName. All rights reserved.
//

import SwiftUI
import shared

class RecipeDetailViewModel: ObservableObject {
    
    private let getRecipe: GetRecipe
    
    @Published var state: RecipeDetailState = RecipeDetailState()
    
    init(recipeId: Int, getRecipe: GetRecipe) {
        self.getRecipe = getRecipe
        onTriggeredEvent(stateEvent: RecipeDetailEvents.GetRecipe(recipeId: Int32(recipeId)))
    }
    
    func onTriggeredEvent(stateEvent: RecipeDetailEvents) {
        switch stateEvent {
        case is RecipeDetailEvents.GetRecipe:
            getRecipe(recipeId: Int((stateEvent as! RecipeDetailEvents.GetRecipe).recipeId))
        case is RecipeDetailEvents.OnRemoveHeadMessageFromQueue:
            doNothing()
        default: doNothing()
        }
    }
    
    private func getRecipe(recipeId: Int) {
        do{
            try self.getRecipe.execute(recipeId: Int32(recipeId))
                .collectCommon(coroutineScope: nil, callback: { dataState in
                    if(dataState != nil) {
                        let data = dataState?.data
                        let message = dataState?.message
                        let loading = dataState?.isLoading ?? false
                        self.updateState(isLoading: loading)
                        
                        if(data != nil) {
                            self.updateState(recipe: data! as Recipe)
                        }
                        if(message != nil) {
                            self.handleMessageByUIComponentType(message!)
                        }
                    } else {
                        print("GetRecipe: DataState is nil")
                    }
                })
        } catch {
            print("Get Recipe: ERROR: \(error)")
        }
    }
    
    private func doNothing() {
        //do nothing
    }
    
    private func updateState(isLoading: Bool? = nil, recipe: Recipe? = nil, queue: Queue<GenericMessageInfo>? = nil) {
        let currentState = (self.state.copy() as! RecipeDetailState)
        self.state = self.state.doCopy(isLoading: isLoading ?? currentState.isLoading, recipe: recipe ?? currentState.recipe, queue: queue ?? currentState.queue)
    }
    
    private func handleMessageByUIComponentType(_ message: GenericMessageInfo) {
        switch message.uiComponentType{
         case UIComponentType.Dialog():
             appendToQueue(message: message)
         case UIComponentType.None():
             print("\(message.description)")
         default:
             doNothing()
         }
    }
    
    private func appendToQueue(message: GenericMessageInfo) {
        let currentState = self.state.copy() as! RecipeDetailState
        let queue = currentState.queue
        let queueUtil = GenericMessageInfoQueueUtil()
        if !queueUtil.doesMessageAlreadyExistInQueue(queue: queue, messageInfo: message) {
            queue.add(element: message)
            updateState(queue: queue)
        }
    }
    
    func removeHeadFromQueue() {
        let currentState = self.state.copy() as! RecipeListState
        let queue = currentState.queue
        do {
            try queue.remove()
            updateState(queue: queue)
        }catch {
            print("\(error)")
        }
    }
}

File: androidFood2Fork/src/main/java/za/co/kernelpanic/food2forkkmm/android/presentation/recipedetail/RecipeDetailViewModel.kt
package za.co.kernelpanic.food2forkkmm.android.presentation.recipedetail

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import za.co.kernelpanic.food2forkkmm.domain.model.GenericMessageInfo
import za.co.kernelpanic.food2forkkmm.domain.model.Recipe
import za.co.kernelpanic.food2forkkmm.domain.model.UIComponentType
import za.co.kernelpanic.food2forkkmm.domain.util.GenericMessageInfoQueueUtil
import za.co.kernelpanic.food2forkkmm.domain.util.Queue
import za.co.kernelpanic.food2forkkmm.interactors.recipe_detail.GetRecipe
import za.co.kernelpanic.food2forkkmm.presentation.recipe_detail.RecipeDetailEvents
import za.co.kernelpanic.food2forkkmm.presentation.recipe_detail.RecipeDetailState
import za.co.kernelpanic.food2forkkmm.presentation.recipe_list.RecipeListState
import java.util.*
import javax.inject.Inject

@HiltViewModel
class RecipeDetailViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val getRecipeUseCase: GetRecipe
) : ViewModel() {

    val state: MutableState<RecipeDetailState> = mutableStateOf(RecipeDetailState())

    init {
        try {
            savedStateHandle.get<Int>("recipeId")?.let { recipeId ->
                onTriggerEvent(RecipeDetailEvents.GetRecipe(recipeId = recipeId))
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun onTriggerEvent(event: RecipeDetailEvents) {
        when (event) {
            is RecipeDetailEvents.GetRecipe -> {
                getRecipe(recipeId = event.recipeId)
            }
            is RecipeDetailEvents.OnRemoveHeadMessageFromQueue -> {
                removeHeadMessage()
            }
            else -> {
                appendToMessageQueue(
                    GenericMessageInfo.Builder()
                        .id(UUID.randomUUID().toString())
                        .title("Error")
                        .uiComponentType(UIComponentType.Dialog)
                        .description("Invalid Event")
                )
            }
        }
    }

    private fun getRecipe(recipeId: Int) {
        getRecipeUseCase.execute(recipeId = recipeId).collectCommon(coroutineScope = viewModelScope) { dataState ->
            state.value = state.value.copy(isLoading = dataState.isLoading)

            dataState.data?.let { recipe ->
                state.value = state.value.copy(recipe = recipe)
            }

            dataState.message?.let { message ->
                appendToMessageQueue(
                    GenericMessageInfo.Builder()
                        .id(UUID.randomUUID().toString())
                        .title("Error")
                        .uiComponentType(UIComponentType.Dialog)
                        .description("Invalid Event")
                )
            }
        }
    }

    private fun appendToMessageQueue(messageInfo: GenericMessageInfo.Builder) {

        if (!GenericMessageInfoQueueUtil().doesMessageAlreadyExistInQueue(
                queue = state.value.queue, messageInfo = messageInfo.build()
            )
        ) {
            val queue = state.value.queue
            queue.add(messageInfo.build())
            state.value = state.value.copy(queue = queue)
        }
    }

    private fun removeHeadMessage() {
        try {
            val queue = state.value.queue
            queue.remove()
            state.value = state.value.copy(queue = Queue(mutableListOf()))
            state.value = state.value.copy(queue = queue)
        }catch (e: Exception) {
            //nothing to remove, queue is empty
        }
    }
}